go/cmd/dolt/commands/add.go

```diff
@@ -119,15 +118,17 @@ func (cmd AddCmd) Exec(ctx context.Context, commandStr string, args []string, dE
 		return 1
 	}
 
-	for {
-		_, err := rowIter.Next(sqlCtx)
-		if err == io.EOF {
-			return 0
-		}
-		if err != nil {
-			cli.PrintErrln(errhand.VerboseErrorFromError(err))
-			return 1
-		}
+	schema, rowIter, err := queryist.Query(sqlCtx, generateSql(apr))
+	if err != nil {
+		cli.PrintErrln(errhand.VerboseErrorFromError(err))
+		return 1
+	}
+
+	_, err = sql.RowIterToRows(sqlCtx, schema, rowIter)
```

Comment:
Is there a better abstraction to ensure we always finish the transaction? Seems like it should be in the deferred close. Ok to ship what works, but I think you'll get tired of this behavior

Reply:
I think I will too. I don't think it's unreasonable that row iterators can have side effects and you have to iterate over all of them to complete the transaction, but it's going to be a pain to have to remember this.

In another language like rust we might be able to make forgetting to close the row iterator a compiler error. But I can't think of anything in go that would make our lives easier here short of making Query return a closefunc (which would be a big refactor but we could do.)

Reply:
I bet there is a way to do this, I just don't know what it is. Maybe a dolt/go pro has a suggestion?

Reply:
So I don't think we want to simply defer the close because closing the rowIter commits the transaction, which we don't want to do if something goes wrong while iterating.

I think the way to do it with a callback, like this:

```
QueryWithCallback(sqlCtx SqlContext, q Queryist, query string, callback func(r Row) error) error {
    schema, rowIter, err := q.Query(sqlCtx, query)
    for _, row := range rowIter {
        err := callback(row)
        if err != nil {
            return err
        }
    }
    rowIter.close()
    return nil
}
```

The trouble is that if our motivation is to make sure that commands don't forget to close the RowIter, then we need to make it impossible for commands to call Queryist.Query directly. We can do this if we make the method private and make all the implementations wrapper classes in the same module as the interface, but that's a bit beyond the scope of this PR.

Thoughts?

Reply:
I think a callback is probably necessary here. Given what you are seeing with the error handling, we may need to pin down the full workflow and all of the call backs used to execute a command end to end. That's on me, but I think we can outline as a group pretty quickly.

Reply:
Yeah, I think this is needed but it's outside the scope of this PR.